File: 12b_solution.rb
require 'set'

input = File.read("./12input.txt").split("\n").map do |line|
  line.split("-")
end

$edges = Hash.new { |h, k| h[k] = Set.new }

input.each do |nodeA, nodeB|
  $edges[nodeA].add nodeB
  $edges[nodeB].add nodeA
end

# returns two booleans: [permission, double_visited]
# - permission determines if path can be taken
# - double_visited indicates if a small cave was visited twice
def check_path(from, to, visited_caves, double_visited)
  if to == 'start'
    [false, double_visited]
  elsif to.downcase == to && visited_caves.include?(to) # visiting small cave twice
    if double_visited
      [false, true]
    else
      [true, true]
    end
  else
    [true, double_visited]
  end
end

# returns number of paths from "from" to end
def find_paths(from, visited_caves, double_visited)
  return 1 if from == 'end'

  $edges[from].inject(0) do |a, neighbour|
    permission, new_double_visited = check_path(from, neighbour, visited_caves, double_visited)

    if !permission
      a
    else
      new_visited_caves = visited_caves + [neighbour]

      a + find_paths(neighbour, new_visited_caves, new_double_visited)
    end
  end
end

paths = find_paths('start', Set.new, false)

puts paths

File: 12b_solution.hs
{-# LANGUAGE OverloadedStrings #-}

import qualified Data.Map as M
import Data.Map (Map)
import qualified Data.Set as S
import Data.Set (Set)
import qualified Data.Text as T
import System.IO

data Node = Start | End | Small Int | Large Int deriving (Eq, Ord)
type NodeSet = Set Node
type EdgeMap = Map Node NodeSet
type NodesToNums = Map T.Text Int

toNode :: T.Text -> NodesToNums -> (Node, NodesToNums)
toNode t nodesToNums
  | t == "start"     = (Start, nodesToNums)
  | t == "end"       = (End, nodesToNums)
  | T.toLower t == t = (Small getNum, updateNums)
  | otherwise        = (Large getNum, updateNums)
  where
    getNum :: Int
    getNum = M.findWithDefault increment t nodesToNums

    updateNums :: NodesToNums
    updateNums
      | M.member t nodesToNums = nodesToNums
      | otherwise              = M.insert t increment nodesToNums

    increment :: Int
    increment = (M.size nodesToNums) + 1

addEdge :: Node -> Node -> EdgeMap -> EdgeMap
addEdge from to edges =
  let edgeSet = M.findWithDefault S.empty from edges
      newEdgeSet = S.insert to edgeSet
   in M.insert from newEdgeSet edges

getEdges :: String -> EdgeMap
getEdges input =
  let inputText = T.pack input
      lines = T.splitOn "\n" inputText
   in fst $ foldr augment (M.empty, M.empty) lines
  where
    augment :: T.Text -> (EdgeMap, NodesToNums) -> (EdgeMap, NodesToNums)
    augment "" a = a
    augment line (edges, nodeMap) =
      let nodes             = T.splitOn "-" line
          (node1, nodeMap1) = toNode (head nodes) nodeMap
          (node2, nodeMap2) = toNode (head (tail nodes)) nodeMap1
          newEdges          = addEdge node1 node2 (addEdge node2 node1 edges)
       in (newEdges, nodeMap2)

type DoubleVisited = Bool
type PathPermitted = Bool

checkPath :: Node -> Node -> NodeSet -> DoubleVisited -> (PathPermitted, DoubleVisited)
checkPath _ Start _ doubleVisited     = (False, doubleVisited)
checkPath _ (Large _) _ doubleVisited = (True, doubleVisited)
checkPath from to visited doubleVisited
  | S.member to visited = ((not doubleVisited), True)
  | otherwise           = (True, doubleVisited)

findPaths :: EdgeMap -> Node -> NodeSet -> DoubleVisited -> Int
findPaths _ End _ _ = 1
findPaths edges from visited doubleVisited =
  let neighbours = M.findWithDefault S.empty from edges
   in foldr count 0 neighbours
  where
    count :: Node -> Int -> Int
    count node acc =
      let (permission, newDoubleVisited) = checkPath from node visited doubleVisited
          newVisited = S.insert node visited
       in if (not permission)
            then acc
            else acc + (findPaths edges node newVisited newDoubleVisited)

findPathsFromStart :: EdgeMap -> Int
findPathsFromStart edges = findPaths edges Start S.empty False

main :: IO ()
main = do
  input <- readFile "12input.txt"
  let edges = getEdges input
  putStrLn $ show $ findPathsFromStart edges
